### DISCUSSION TITLE
preemtive look rotation

### DISCUSSION THREAD

**Discussion by kianzarrin** [Asu4ni/CitiesSkylines-FPSCamera#42] [Ideas]
normally ppl will look where they go. if they turn right first they look right and then they turn not the way around.
It would be nice for this tool to do the same.
it should be as easy as getting direction from future frames.

**Comment by kianzarrin** [comment]
BTW there is a TMPE diff that can show future stuff:
![image](https://user-images.githubusercontent.com/26344691/170148926-979a3f4f-6177-4867-adee-d331f5d81d5b.png)
purple = vehicle segment (use for collision detection maybe)
circles = targetpos0-3
I think we should look toward the vehicle segment.
or we can put a preemptive turn slider to let the user choose.

**Reply by kianzarrin** [reply]
credit: krzychu: 
[RenderFrameOverlay.zip](https://github.com/Asu4ni/CitiesSkylines-FPSCamera/files/8766876/RenderFrameOverlay.zip)

**Comment by kianzarrin** [comment]
nailed it!
https://user-images.githubusercontent.com/26344691/170259211-ed41413d-a4e6-4cf9-9e89-f5c624404152.mp4

**Reply by kianzarrin** [reply]
OK now we need to discuss about the UI options and how do the PR.
This is the code:
```cs
public Positioning GetPositioning()
{
	_vehicle.GetSmoothPosition(_vid._index, out var position, out var rotation1);
	var rotation2 = Quaternion.LookRotation(GetSmoothLookPos() - position);
	const float rotationFactor = 0.9f;
	var rotation = Quaternion.Lerp(rotation1, rotation2, rotationFactor);
	return new Positioning(Position._FromVec(position), Angle._FromQuat(rotation));
}

Vector3 _lastLookPos;
Vector3 _prevTargetPos;
public Vector3 GetSmoothLookPos()
{
	Vector3 pos = _vehicle.m_targetPos3;
	bool targetPosChanged = _prevTargetPos != pos;
	if (targetPosChanged) {
		_lastLookPos = _prevTargetPos;
		_prevTargetPos = pos;
	}

	// TODO use slider to decide which frame/targetPos to use.
	return Vector3.Lerp(_lastLookPos, pos, GetBezierOffset());
}
```

**Comment by Asu4ni** [comment]
@kianzarrin, really appreciate your effort!! The video looks neat.

In case I misunderstand your intention, let me repeat your goal: you'd like to make camera rotate a bit earlier according to the target's future position, eh?

If that's the case, I'd assume we'd have to:
- a new method to get future position for `CSkyL.Game.Object.IObjectToFollow`. e.g. `Vehicle.GetFuturePositioning`
- a new option (maybe a slider) of the degree of rotating earlier (value 0 would be the status quo)
- `FPSCamera.Cam.FollowCam.GetPositioning` should linearly interpolate (according to the new option) between positionings from `GetPositioning` and `GetFuturePositioning`.

In addition, I'd like to say that I'm new to the modding community. Hence, I'm not quite familiar to a lot of the details. `_vehicle.m_targetPos3`, for example, is something I just learned from your comment. I'd be glad to have others help to improve the project, but please bear with me having basic questions here and there.

**Reply by Asu4ni** [reply]
I've created issues for this #43 & Asu4ni/CSkyL#1

**Comment by kianzarrin** [comment]
Hi @Asu4ni , 
I couldn't reach you on discord. I am trying to write some TMPE code which gives me more insight into path calculations.
armed with this new knowledge, I think I can improve look ahead look rotation in a way that fits into previous look ahead framework. are you interested?

PS: what happened to that user issue that was reported?
